altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+                self.head = unsafe { (*self.head).next_block };
+            }
+            else {
+                unsafe { (*previous).next_block = (*current).next_block; }
+            }
+        }
+        // Current block is larger than required and has the correct alignment
+        else if current_size > using_size && align_diff == 0 {
+            unsafe { (*current).block_size -= using_size; }
+            if self.head == current {
+                self.head = self.shift_block_forward(current, using_size);
+            }
+            else {
+                unsafe {
+                    (*previous).next_block = self.shift_block_forward(current, using_size);
+                }
+            }
+        }
+        // Current block is larger than required but has exactly the right size to
+        // accomodate the alignment and size requirements
+        else if current_size == using_size + align_diff {
+            unsafe { (*current).block_size = align_diff; }
+        }
+        // Current block is larger than required and because of alignment, the allocation
+        // divides it in two.
+        else {
+            unsafe {
+                (*current).block_size = current_size - using_size - align_diff;
+                let upper_block = self.shift_block_forward(current, using_size + align_diff);
+                ptr::write(current as *mut BlockHeader, BlockHeader::new(align_diff));
+                (*current).next_block = upper_block;
+            }
+        }
+        alloc_start as *mut u8
+    }
+
+    /*
+    TODO: We can still implement merging to deal with fragmentation
+    - Nothing adjacent: Make new block, connect to closest blocks
+    - Adjacent at tail: Merge with tail block, move block, switch leading ptr
+    - Adjacent at lead: Merge with lead, no additional changes
+    - Adjacent at both: Merge two with lead (add sizes, switch lead ptr)
+    */
+
+    // Deallocates memory, placing it back in the free list for later use
+    // Adds a free block to the list based on alloc_ptr so that the list remains
+    // sorted based on memory position.
+    pub fn deallocate(&mut self, alloc_ptr: *mut u8, size: usize, _align: usize) {
+        // We can immediately add the block at the deallocated position
+        let alloc_block_ptr = alloc_ptr as *mut BlockHeader;
+        let used_memory = alignment::align_up(size, self.block_hdr_size);
+        unsafe {
+            ptr::write(&mut *alloc_block_ptr, BlockHeader::new(used_memory));
+        }
+
+        // Memory location to be added at head of the list if the list is empty
+        // or if it's position in memory comes before the current head.
+        if self.head.is_null() || alloc_block_ptr < self.head {
+            unsafe { (*alloc_block_ptr).next_block = self.head; }
+            self.head = alloc_block_ptr;
+            return;
+        }
+        let (mut previous, current) = {
+            let free_mem_between = |previous, current| {
```

Comment:
This closure can be used inline with the `find_block` call if the signature of `find_block` is changed as suggested.